### PR TITLE
Emit dense LM-head logits directly in FP32 on CUDA

### PR DESCRIPTION
**Pull request by paulzhang-tm**
Use torch.mm with out_dtype=torch.float32 for unquantized CUDA LM heads, including RL-on-policy tied weights. Preserve the existing matmul fallback for non-CUDA devices and avoid recasting logits that are already FP32.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #30076153859](https://github.com/sgl-project/sglang/actions/runs/30076153859)
Latest PR Test (Extra): :x: [Run #30076153578](https://github.com/sgl-project/sglang/actions/runs/30076153578)

### CODE REVIEW

**Comment by sshleifer** [comment]
LGTM TY!

**Review by guapisolo — APPROVED** [review]
Approved by Yiming and me, need code owner's review

**Comment by paulzhang-tm** [comment]
Logprobs changing result in CI failures, though this is strictly more precision. Thoughts on how to proceed?

**Comment by ispobock** [comment]
Some CIs cannot pass:
https://github.com/sgl-project/sglang/actions/runs/30076153859/job/89684332480#step:13:3115 
https://github.com/sgl-project/sglang/actions/runs/30076153859/job/89684332397#step:13:1750
